aggregator/src/cache.rs

```diff
@@ -206,3 +210,264 @@ impl PeerAggregatorCache {
+    task_aggregator: Option<Arc<TaskAggregator<C>>>,
+}
+
+pub const TASK_AGGREGATOR_CACHE_DEFAULT_TTL: Duration = Duration::from_seconds(600);
+
+impl<C: Clock> TaskAggregatorCache<C> {
+    pub fn new(
+        clock: C,
+        datastore: Arc<Datastore<C>>,
+        report_writer: ReportWriteBatcher<C>,
+        taskprov_enabled: bool,
+        ttl: Duration,
+    ) -> Self {
+        let ttl = ttl.as_seconds();
+        let jitter = (ttl as f64 * 0.1) as u64;
+        Self {
+            clock,
+            datastore,
+            report_writer: Arc::new(report_writer),
+            entries: Default::default(),
+            taskprov_enabled,
+            ttl_seconds: Uniform::new(ttl - jitter, ttl + jitter),
+        }
+    }
+
+    /// Calculates the next possible cache TTL. This contains jitter to prevent all tasks in the
+    /// cache from falling on the same deadline on application startup.
+    fn next_ttl(&self) -> Time {
+        self.clock.now().saturating_add(&Duration::from_seconds(
+            thread_rng().sample(self.ttl_seconds),
+        ))
+    }
+
+    pub async fn get(&self, task_id: &TaskId) -> Result<Option<Arc<TaskAggregator<C>>>, Error> {
+        // Step one: grab the existing entry for this task, if one exists. If there is no existing
+        // entry, write a new (empty) entry.
+        let entry = {
+            // Unwrap safety: mutex poisoning.
+            let mut entries = self.entries.lock().unwrap();
+            Arc::clone(entries.entry(*task_id).or_insert(Arc::new(Mutex::new(
+                TaskAggregatorCacheEntry {
+                    expiration: self.next_ttl(),
+                    is_initialized: false,
+                    task_aggregator: None,
+                },
+            ))))
+        };
+
+        // Step two: if the entry is empty, fill it via a database query. Concurrent callers
+        // requesting the same task will contend over this lock while awaiting the result of the DB
+        // query, ensuring that in the common case only a single query will be made for each task.
+        let task_aggregator = {
+            let mut entry = entry.lock().await;
+            if self.clock.now() > entry.expiration || !entry.is_initialized {
+                *entry = TaskAggregatorCacheEntry {
+                    expiration: self.next_ttl(),
+                    is_initialized: true,
+                    task_aggregator: self
+                        .datastore
+                        .run_tx("task_aggregator_get_task", |tx| {
+                            let task_id = *task_id;
+                            Box::pin(async move { tx.get_aggregator_task(&task_id).await })
+                        })
+                        .await?
+                        .map(|task| TaskAggregator::new(task, Arc::clone(&self.report_writer)))
+                        .transpose()?
+                        .map(Arc::new),
+                };
+            }
+            entry.task_aggregator.clone()
```

Comment:
nit: prefer `Arc::clone(&entry.task_aggregator)` in order to make it clear we're cloning an `Arc` (cheap) instead of a `TaskAggregator` (expensive)

Reply:
I wasn't able to take this suggestion because the `task_aggregator` is now an `Option<Arc<_>>`. I can't `Arc::clone()` an `Option<Arc<_>>`, I have to `.clone()` it so `Option::clone()` can call the underlying clone.